### ISSUE TITLE
ClassCastException on maps with UUIDs as key values

### ISSUE THREAD

**Issue by Foorcee** [MorphiaOrg/morphia#1625]
When we want to load objects with our UuidAsStringCodec codec, there are string instances in maps. Basically the saving and loading of lists works.

https://github.com/Cytooxien/reproducer_morphia_map

Test output:
```
entity.equals(entity1) = false
class java.util.UUID

java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.UUID (java.lang.String and java.util.UUID are in module java.base of loader 'bootstrap')

	at java.base/java.util.HashMap.forEach(HashMap.java:1425)
	at com.foo.ReproducerTest.reproduce(ReproducerTest.java:72)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:221)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
```

** Please complete the following information: **
    Server Version: 4.0.16
    Driver Version: ...
    Morphia Version: ...

**Comment by evanchooly** [comment]
I think I know roughly the why and what of this but I'm away for a week and don't have my hardware handy to confirm it.  When I get back home in a week or so, I'll take a look.

**Comment by evanchooly** [comment]
note to self: problem same fix as #1622

**Comment by evanchooly** [comment]
OK.  So there are a couple of things going on here.

1. Your codec simply can't work.  The way that property is defined, it's going to end up getting mapped to a `MapCodec` from deep withing the bowels of Morphia.  That's all fine and good and as expected.  However, I know you're thinking that you've registered a codec for UUIDs that when Morphia goes to read those key values, it should use your codec because those keys are UUIDs.  But sadly that's not quite how it works.  MongoDB itself only allows for String key types and so Morphia does some magic (see [2] below) to convert keys to/from String values.  Because of the readers work in the driver, we can't simply delegate to a different codec to read the keys because it's in a `name` state and so has to be read in a specific way.  Codecs typically read data while in a `value` state.  Now, if your *values* were UUIDs, your codec should be called to active duty to translate the data from the database directly.  So all that to say, is that we're pulling the keys with a `name` read and that is necessarily a String coming from the database.  This String must then be converted to a UUID before being put in the Map and it's *this* part that is missing.
2. Morphia uses what are creatively called `Conversions` to make this mapping to/from Strings and other types.  For UUID, this conversion is missing.  I'll use this issue to track adding that conversion so that Morphia handles it for you and then you won't need your codec at all because Morphia will magically know what to do.

Hopefully this all makes sense.  There's not much left for 2.2.2 (a few clean ups and questions to resolve, I think) and then I can push out a release that should get you squared away.  Does this all make sense?
